### ISSUE TITLE
[Experiment] PSEM trainable formulation gate — frozen/trainable × direct/structured

### ISSUE THREAD

**Issue by kapitalismho** [kapitalismho/PuriPuly-heart#72]
# [Experiment] PSEM trainable formulation gate — frozen/trainable × direct/structured

## Status / authority

**Small exploratory training gate only.**

This issue is intentionally narrower than the PSEM training plan. It does **not** choose the final fine-tuning recipe, model family, distillation strategy, compact architecture, multilingual data program, or production runtime.

Primary branch:

```text
experiment-v2-speaker-change-turn-boundaries-ls
```

The single question is:

> **With one existing pretrained speech encoder, does allowing task-specific encoder adaptation and/or training a structured local speaker-state target improve the natural continuous low-false-event speaker-change frontier?**

The result decides which learning direction is worth pursuing next. It is not a release-readiness claim.

---

# Read first

Use the existing branch as the implementation/evidence base rather than creating a parallel research stack.

Read at minimum:

```text
AGENTS.md
experiments/speaker_representation_scd/models/source_registry.json
experiments/speaker_representation_scd/R7B_FIXED_LAG_LOCAL_SEGMENTATION_ADDENDUM.en.md
experiments/speaker_representation_scd/configs/r7b/fixed_lag_local_segmentation.json
experiments/speaker_representation_scd/r7b_local_segmentation.py
experiments/speaker_turn_boundary/README.md
```

R7-B is the conceptual starting point for the structured arm. Existing speaker-turn-boundary code is the starting point for canonical timeline / event matching / false-event accounting where compatible.

---

# One encoder only

Use the already pinned branch artifact:

```text
wavlm-base-plus
```

from:

```text
experiments/speaker_representation_scd/models/source_registry.json
```

Do not turn this issue into a WavLM vs UniSpeech vs ERes / ECAPA benchmark.

Do not change the pretrained checkpoint identity after seeing A/B/C/D results.

---

# Primary 2 × 2 matrix

Run exactly these four primary arms:

| ID | Encoder | Target | Question |
|---|---|---|---|
| `A-FROZEN-DIRECT` | WavLM frozen | direct speaker-change event | frozen/direct control |
| `B-TRAINABLE-DIRECT` | WavLM trainable/adapted | direct speaker-change event | does task adaptation alone help? |
| `C-FROZEN-STATE` | WavLM frozen | structured local speaker state/relation | does target formulation alone help? |
| `D-TRAINABLE-STATE` | WavLM trainable/adapted | structured local speaker state/relation | primary hypothesis arm |

Conceptually:

```text
                     Direct change       Structured state

Frozen encoder             A                    C

Trainable encoder          B                    D
```

The purpose is to make these comparisons interpretable:

```text
A -> B : effect of allowing encoder adaptation under direct supervision
A -> C : effect of structured-state supervision with frozen evidence
B -> D : extra value of structured-state supervision after adaptation
C -> D : effect of encoder adaptation under structured supervision
```

---

# Keep everything else common

Across A/B/C/D, keep the following the same unless an implementation constraint makes exact sharing impossible:

```text
pretrained checkpoint identity
source audio
train/dev fold identities
temporal context / availability contract
GT normalization
final speaker-change event semantics
event matching / duplicate handling
evaluation code
comparable task-head capacity
```

Any unavoidable difference must be documented before interpreting the comparison.

Do not change data, context, checkpoint, event semantics, and training formulation simultaneously to rescue a weak arm.

---

# Frozen vs trainable

## Frozen arms: A / C

The WavLM encoder does not receive gradients. Train only the task-specific head/state layer required by the arm.

## Trainable arms: B / D

The WavLM encoder must receive task-loss gradients through one **reasonable bounded adaptation recipe**.

The exact fine-tuning method is **not a research variable in this issue**.

Before the official B/D runs:

1. choose one reasonable bounded adaptation recipe;
2. record which encoder parameters are trainable and the basic training recipe;
3. use the **same encoder adaptation recipe for B and D**;
4. do not tune B and D independently.

Adapter vs partial fine-tuning, exact trainable layer count, optimizer, learning rate, batch size, and similar implementation details are delegated to the experiment implementation unless a blocker requires escalation.

Do not run a fine-tuning-method sweep.

---

# Direct target

The direct arms ask the model to predict the final local speaker-change / handoff event directly from the available evidence.

Conceptually:

```text
audio -> WavLM -> temporal/task head -> P(speaker_change)
```

Use one canonical speaker-change event definition for both direct and structured arms.

The direct arm is intentionally simple. Do not add a second hidden structured speaker-state system to the direct control.

---

# Structured target

The structured arms should stay close to the already-tested **R7-B local-state formulation** rather than inventing a new diarization model.

Required core supervision:

```text
1. local speech state
   silence / singleton / overlap

2. local speaker relation
   same speaker / different speaker between reliable singleton regions
```

Then derive the same final speaker-change event used to evaluate the direct arms from the learned state/relation process.

Conceptually:

```text
audio
  -> WavLM
  -> local state / same-speaker relation
  -> local state transition
  -> speaker-change event
```

The goal is to test whether denser local state/relation supervision is easier to learn and more robust at low false-event rates than a sparse direct change target.

Do **not** expand this gate into:

```text
full diarization
global speaker IDs
long-term clustering
persistent speaker registry
speaker attractor research
new current-speaker/other-speaker target families
large state-machine design search
```

Those may be follow-up work if D is promising.

---

# Data

Reuse the existing **natural continuous AMI / AliMeeting-style material already used by the speaker research branch** where GT and continuous evaluation are available.

This gate does not require a new multilingual/product corpus acquisition program.

Required properties for the material used in A/B/C/D:

```text
continuous natural speech
speaker/activity GT
real speaker changes
long same-speaker negative regions
some overlap / short-turn behavior where available
```

Use the same train/dev source identities for all four arms.

Existing R7-B fold/context material is the preferred starting point. It may be extended with additional existing AMI/AliMeeting training material if needed for trainability, but the resulting split must be frozen before official A/B/C/D comparison and shared by all arms.

Treat this experiment as **development-known / direction-selection evidence** unless a genuinely untouched natural panel is separately frozen before training.

Do not claim KO/JA/code-switch/product-domain generalization from this gate.

---

# Temporal / event contract

Reuse the existing R7-B / speaker-turn-boundary research conventions where practical rather than opening a new latency study.

For this issue, the only hard requirement is:

> **A/B/C/D must see the same effective temporal evidence and be evaluated against the same event semantics.**

Do not make 80 ms vs 100 ms grid, 500 ms vs 1 s context, or another timing geometry an experimental axis here.

If the implementation reuses the current R7-B geometry directly, that is acceptable and preferred for simplicity.

---

# Training data sampling

Official training should expose the model to both:

```text
event-centered / event-balanced examples
and
natural source-time-continuous examples
```

The first makes rare speaker changes learnable; the second exposes the model to the realistic negative base rate required by the low-FP evaluation.

Exact batch composition is an implementation choice, but an official run that contains only boundary-centered examples is not sufficient for this gate.

No teacher / pseudo-label / KD target is used in A/B/C/D.

---

# Evaluation

## Headline result

The headline comparison is:

> **Natural continuous speaker-change recall vs false events per source hour.**

Report A/B/C/D on the same frontier.

At minimum include reference rows where the available negative exposure supports them:

```text
1 FE/h
5 FE/h
10 FE/h
20 FE/h
```

Do not over-interpret unsupported ultra-low-FP rows when the evaluation duration is too short.

Also report the existing event-tolerance metrics needed to understand timing, preferably using the branch's established ±250 ms / ±500 ms conventions.

## Diagnostics

For the structured arms, also report enough state metrics to verify that training is doing what was intended:

```text
silence / singleton / overlap quality
same-speaker / different-speaker relation quality
```

These are diagnostics. The final comparison remains the natural continuous speaker-change frontier.

---

# Suggested execution order

Run the cheap/known controls first:

```text
1. A-FROZEN-DIRECT
2. C-FROZEN-STATE
```

Use them to verify the new/common harness is broadly consistent with prior frozen-representation evidence.

Then run:

```text
3. B-TRAINABLE-DIRECT
4. D-TRAINABLE-STATE
```

A/C are not pass/fail gates for starting B/D unless a correctness problem is found. Their purpose is to make the adaptation comparison trustworthy.

---

# Interpretation

## If A -> B improves strongly

Task-specific encoder adaptation matters. The previous frozen representation was a major bottleneck.

## If A -> C improves strongly

The target/formulation matters even without changing the representation.

## If B -> D and C -> D improve strongly

The combination of trainable representation + structured state is the most promising direction.

This is the main hypothesis behind the experiment.

## If B is already as good as or better than D

Adaptation appears sufficient; do not add structured-state complexity without measured benefit.

## If C is already as good as D

A frozen representation plus better target/state formulation may be sufficient.

## If none of A/B/C/D improves the natural low-FP frontier

Stop this gate.

Do not respond by automatically sweeping larger heads, more backbones, longer contexts, KD, or scratch PSEM architectures.

Report the failure and inspect data coverage, labels, target observability, and domain mismatch as the next research question.

This is an exploratory direction gate; a visibly better natural continuous frontier is more important than manufacturing a complicated significance threshold for this first comparison.

---

# Explicitly out of scope

```text
pretrained-backbone leaderboard
fine-tuning-method comparison / LoRA sweep
scratch/random-init PSEM comparison
teacher / KD / pseudo-labeling
5M / 2M / 1M compression
INT8 / CPU deployment optimization
new multilingual corpus program
KO / JA / code-switch release claims
product wiring / UX experiment
full diarization
persistent/global speaker identity
architecture sweep
context/latency sweep
language auxiliary experiments
```

Success here means only:

> **We have evidence for which of frozen/trainable representation and direct/structured target formulation is worth carrying into the next PSEM training stage.**

---

# Deliverables

Create one compact experiment area or extend the existing speaker-representation experiment area without duplicating canonical infrastructure.

Required outputs:

```text
- frozen experiment config / receipt
- exact A/B/C/D model + training identities
- exact train/dev source identities
- raw prediction artifacts needed to reproduce event scoring
- A/B/C/D natural continuous recall-FE/h table/curve
- structured-state diagnostic table
- representative high-scoring false-event examples
- concise RESULTS.md with one final interpretation
```

The final report must answer:

```text
Did adaptation help?          A -> B and C -> D
Did structured targets help? A -> C and B -> D
Which arm is the best next-direction candidate?
```

---

# Definition of Done

- [ ] Existing pinned `wavlm-base-plus` identity is used for all four arms.
- [ ] One shared data/split/context/event contract is frozen before official comparison.
- [ ] `A-FROZEN-DIRECT` completed.
- [ ] `B-TRAINABLE-DIRECT` completed.
- [ ] `C-FROZEN-STATE` completed.
- [ ] `D-TRAINABLE-STATE` completed.
- [ ] B and D use the same encoder adaptation recipe.
- [ ] Direct and structured arms are scored against the same final speaker-change event semantics.
- [ ] Structured arms implement the bounded R7-B-style core: speech state + local same/different-speaker relation.
- [ ] Natural continuous recall vs FE/h is reported for all four arms.
- [ ] Structured-state diagnostic metrics are reported.
- [ ] No teacher/KD, scratch-model, backbone, architecture, or fine-tuning-method sweep is introduced.
- [ ] The report explicitly states what A->B, A->C, B->D, and C->D imply.
- [ ] One next research direction is recommended, or the gate is closed as unsuccessful.

---

# Prior evidence / implementation references

```text
experiments/speaker_representation_scd/R7B_FIXED_LAG_LOCAL_SEGMENTATION_ADDENDUM.en.md
experiments/speaker_representation_scd/configs/r7b/fixed_lag_local_segmentation.json
experiments/speaker_representation_scd/r7b_local_segmentation.py
experiments/speaker_representation_scd/models/source_registry.json
experiments/speaker_turn_boundary/README.md
```

Related experiment context:

- #58 — teacher/reference characterization; independent of this gate
- prior R5/R7/R7-B/R8/R9 results under `experiments/speaker_representation_scd/`

Those experiments are context and reusable implementation evidence. Their old pass/fail thresholds are not automatically inherited as the success gate for this issue.

**Comment by kapitalismho** [comment]
## Experiment stopped and final summary

Per the decision made on 2026-08-17, this experiment is stopped here. No further regeneration, feature extraction, training, adapter/structured modifications, or follow-up error analysis will be performed.

### Most important scope correction

This implementation **did not perform the actual encoder adaptation/fine-tuning described in the issue.**

- The pretrained encoder parameters were frozen in every arm.
- `B-TRAINABLE-DIRECT` and `D-TRAINABLE-STATE` trained only a **64-dimensional residual bottleneck adapter + task head after the encoder output**, not the encoder itself.
- pretrained encoder parameters trainable = `0`
- The scratch/random-init arm was not run.

Therefore, the results below **do not determine fine-tuning vs. scratch.** In addition, because B/D did not satisfy the issue's contract that “encoder must receive task-loss gradients,” the Definition of Done for evaluating encoder adaptation effects cannot be considered satisfied either.

### What was actually compared

Per the owner amendment, the same four arms were applied to the following three pretrained embeddings.

- `eres2netv2-standard-prepool`
- `wavlm-base-plus`
- `mhubert-147`

| Arm | Actual meaning |
| --- | --- |
| A | Frozen pretrained embedding + direct event head |
| B | Frozen pretrained embedding + output adapter + direct event head |
| C | Frozen pretrained embedding + structured state/relation head + event projection |
| D | Frozen pretrained embedding + output adapter + structured head + event projection |

All 12 arms used the same 10 natural continuous meetings, 5-fold OOF split, context, event semantics, duplicate suppression, and one-to-one matching.

### Evaluation correction

The initial interpretation centered on 10 FE/h + 250 ms overvalued thresholds with high precision but extremely low recall. The final reevaluation was corrected to:

- use the full score range
- evaluate 100/250/500 ms collars in parallel
- use the threshold that maximizes the mean F1 across the three collars
- treat FE/h only as a reference annotation, not as a policy or threshold constraint

### 12-arm provisional results

The values below are the average F1 across the 100/250/500 ms collars (Macro F1).

| Model | A: Frozen Direct | B: Output-adapter Direct | C: Frozen Structured | D: Output-adapter Structured |
| --- | ---: | ---: | ---: | ---: |
| ERes2NetV2 | 37.34% | **38.32%** | 28.88% | 28.89% |
| WavLM Base+ | **41.59%** | 41.31% | 30.23% | 30.51% |
| mHuBERT-147 | **44.31%** | 44.30% | 34.38% | 34.18% |

Within the tested **frozen pretrained-feature pipelines**, the best result was `mhubert-147 / A-FROZEN-DIRECT`.

mHuBERT-A:

| Collar | Precision | Recall | F1 |
| ---: | ---: | ---: | ---: |
| 100 ms | 29.00% | 44.21% | 35.02% |
| 250 ms | 38.21% | 58.26% | 46.15% |
| 500 ms | 42.84% | 65.32% | 51.75% |

### Observed conclusions

1. The current **post-encoder output adapter** did not produce a consistent improvement.
   - This does not mean that encoder fine-tuning is ineffective.
2. The current R7-B-style **end-to-end structured-to-event pipeline** was about 8–11 percentage points lower than the direct head across all three models.
   - Since state macro-F1 was 69.61–73.69% and relation balanced accuracy was 74.47–83.81%, this cannot be taken as evidence that the structured representation itself failed.
   - State/relation training and event projection losses are confounded.
3. mHuBERT-A errors include both timing issues and remote false activations.
   - 1,040 FPs at 100 ms become TPs at 500 ms.
   - For 500 ms TPs, the median absolute timing error is 60 ms and p90 is 330 ms.
   - Of 4,025 FPs at 500 ms, 3,845 (95.5%) are remote FPs more than 500 ms away from every GT.
   - Of the remote FPs, 2,744 (71.4%) are continuous same-speaker singletons, and 711 (18.5%) are overlap continuations.
   - Threshold-independent post-NMS candidate coverage is 99.29% at 500 ms (missing 33/4,619).

In other words, rather than an absence of candidate geometry, what was observed was **localization error + high false activation on same-speaker/ongoing-overlap cases**, with the latter being larger in terms of FP count.

### Evidence limitations and stopped state

- **Development-known provisional evidence** from reevaluating existing OOF raw predictions with the corrected full-range evaluator
- The 12 arms were not officially regenerated from a clean commit of the corrected final code.
- Independent review is incomplete.
- There is no evidence from an untouched test panel, the product domain, or multilingual generalization.
- Because there are no laughter/prosody annotations, remote FPs were not semantically classified for those categories.
- No context/latency sweep was performed.
- 100/250/500 ms are matching collars, not a product latency policy.

### Final one-sentence summary

> Among comparisons based on frozen pretrained embeddings, the mHuBERT direct head performed best, and the current output adapter and structured-to-event implementation did not improve performance. However, because actual encoder fine-tuning and scratch were not tested, the original fine-tuning-vs.-scratch decision remains unresolved.

The issue will retain only this conclusion record; its current status will not be changed.
